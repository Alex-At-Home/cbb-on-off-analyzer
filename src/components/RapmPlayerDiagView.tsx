// React imports:
import React, { useState } from 'react';

import _ from "lodash";

// Next imports:
import { NextPage } from 'next';

// Utils
import { RapmInfo, RapmPlayerContext, RapmPreProcDiagnostics, RapmProcessingInputs, RapmUtils } from "../utils/stats/RapmUtils";

type Props = {
  rapmInfo: RapmInfo,
  player: Record<string, any>,
  globalRef: React.RefObject<HTMLDivElement>
};

const RapmPlayerDiagView: React.FunctionComponent<Props> = (({rapmInfo, player, globalRef}) => {
  try {
    const ctx = rapmInfo.ctx;
    const offWeights = rapmInfo.offWeights.valueOf();
    const offInputs = rapmInfo.offInputs;
    const defInputs = rapmInfo.defInputs;

    const gotoGlobalDiags = () => {
      if (globalRef.current) {
        globalRef.current.scrollIntoView({
          behavior: 'smooth',
          block: 'start',
        });
      }
    };

    const rapmOff = (player.rapm?.off_adj_ppp?.value || 0);
    const rapmDef = (player.rapm?.def_adj_ppp?.value || 0);
    const calcPriorDiags = () => {
      if (ctx.unbiasWeight > 0) {
        const totalOffPoss = ctx.teamInfo?.off_poss?.value;
        const teamOffAdj = ((ctx.teamInfo.off_adj_ppp?.value || ctx.avgEfficiency) - ctx.avgEfficiency);
        const teamDefAdj = ((ctx.teamInfo.def_adj_ppp?.value || ctx.avgEfficiency) - ctx.avgEfficiency);

        const col = ctx.playerToCol[player.playerId];
        const offPossPcts = offWeights[ctx.numLineups]; //(use as approx for def also)
        const offPossPctStr = (100.0*(offPossPcts[col]/ctx.unbiasWeight)).toFixed(0);

        const [ sigmaRapmOff, sigmaRapmDef ] = rapmInfo?.noUnbiasWeightsDiags?.map((vec) =>
          _.reduce(vec, (acc, n: number, i: number) => acc + n*offPossPcts[i]/ctx.unbiasWeight) || 0
        ) || [0, 0];
        const offPriorTotalDiff = teamOffAdj - sigmaRapmOff;
        const defPriorTotalDiff = teamDefAdj - sigmaRapmDef;

        const [ offUnbiasRapm, defUnbiasRapm ] = rapmInfo?.noUnbiasWeightsDiags?.map((vec) => vec[col]) || [0, 0];
        const offPrior = rapmOff - offUnbiasRapm;
        const defPrior = rapmDef - defUnbiasRapm;

        return [
          offPrior, defPrior,
          <ul>
            <li>We calculate the total team priors (off=[<b>{offPriorTotalDiff.toFixed(2)}</b>] def=[<b>{defPriorTotalDiff.toFixed(2)}</b>]) from
            the reduction in the team adjusted efficiency due to the regression factor, eg compare <em>observed</em> (off=[<b>{teamOffAdj.toFixed(2)}</b>] def=[<b>{teamDefAdj.toFixed(2)}</b>])
            vs <em>derived from RAPM</em> (off=[<b>{sigmaRapmOff.toFixed(2)}</b>] def=[<b>{sigmaRapmDef.toFixed(2)}</b>]).
            </li>
            <li>Then we calculate a player's contribution to the team prior - currently this is mostly based on their % on floor, [<b>{offPossPctStr}%</b>] (of [<b>{totalOffPoss}</b>] poss);
            eg for the offense approximately [<b>{offPossPctStr}%</b>] * (1/5) * [<b>{offPriorTotalDiff.toFixed(2)}</b>] = [<b>{(0.2*offPossPcts[col]*offPriorTotalDiff/ctx.unbiasWeight).toFixed(2)}</b>],
            plus some smaller lineup adjustments.
            </li>
          </ul>
        ];

      } else {
        return [ 0, 0, <ul><li>Not currently supported</li></ul> ]; //(not currently supported)
      }
    };
    const [ priorOff, priorDef, priorDiagListEl ] = calcPriorDiags() as [ number, number, Element ];
    const totalPrior = priorOff - priorDef;
    const rawRapmOff = rapmOff - priorOff;
    const rawRapmDef = rapmDef - priorDef;
    const totalRawRapm = rawRapmOff - rawRapmDef;

    return <span>
        <b>RAPM diagnostics for [{player.playerId}]:</b> adj_off=[<b>{rapmOff.toFixed(2)}</b>], adj_def=[<b>{rapmDef.toFixed(2)}</b>]
        <ul>
          <li>RAPM contribution: off=[<b>{rawRapmOff.toFixed(2)}</b>], def=[<b>{rawRapmDef.toFixed(2)}</b>], total=[<b>{totalRawRapm.toFixed(2)}</b>]</li>
          <li>Priors contribution: off=[<b>{priorOff.toFixed(2)}</b>], def=[<b>{priorDef.toFixed(2)}</b>], total=[<b>{totalPrior.toFixed(2)}</b>]</li>
            {priorDiagListEl}
        </ul>
        (<b>More player diagnostics to come...</b>)<br/>(<a href="#" onClick={(event) => { event.preventDefault(); gotoGlobalDiags() }}>Scroll to global RAPM diagnostics</a>)
      </span>;
    }
    catch (err) { //Temp issue during reprocessing
      return <span>Recalculating diags, pending {err.message}</span>;
    }
});

export default RapmPlayerDiagView;
