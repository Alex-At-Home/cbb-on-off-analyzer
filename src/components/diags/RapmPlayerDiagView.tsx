// React imports:
import React, { useState } from 'react';

import _ from "lodash";

// Next imports:
import { NextPage } from 'next';

// Utils
import { RapmInfo, RapmPlayerContext, RapmPreProcDiagnostics, RapmProcessingInputs, RapmUtils } from "../../utils/stats/RapmUtils";

type Props = {
  rapmInfo: RapmInfo,
  player: Record<string, any>,
  globalRef: React.RefObject<HTMLDivElement>
};

const RapmPlayerDiagView: React.FunctionComponent<Props> = (({rapmInfo, player, globalRef}) => {
  try {
    const ctx = rapmInfo.ctx;
    const offWeights = rapmInfo.offWeights.valueOf();
    const offInputs = rapmInfo.offInputs;
    const defInputs = rapmInfo.defInputs;

    const gotoGlobalDiags = () => {
      if (globalRef.current) {
        globalRef.current.scrollIntoView({
          behavior: 'smooth',
          block: 'start',
        });
      }
    };

    const rapmOff = (player.rapm?.off_adj_ppp?.value || 0);
    const rapmDef = (player.rapm?.def_adj_ppp?.value || 0);

    const col = ctx.playerToCol[player.playerId];

    const totalOffPoss = ctx.teamInfo?.off_poss?.value;
    const teamOffAdj = ((ctx.teamInfo.off_adj_ppp?.value || ctx.avgEfficiency) - ctx.avgEfficiency);
    const teamDefAdj = ((ctx.teamInfo.def_adj_ppp?.value || ctx.avgEfficiency) - ctx.avgEfficiency);

    const offPoss = offInputs.playerPossPcts[col]!;
    const defPoss = defInputs.playerPossPcts[col]!;
    const offPossPctStr = (100.0*(offInputs.playerPossPcts[col]!)).toFixed(0);

    // Prior (luck adjusted o/drtg)
    const offPrior = ctx.priorInfo.playersWeak?.[col]?.off_adj_ppp || 0;
    const defPrior = ctx.priorInfo.playersWeak?.[col]?.def_adj_ppp || 0;

    const offUnbiasRapm = offInputs.rapmRawAdjPpp[col];
    const defUnbiasRapm = defInputs.rapmRawAdjPpp[col];

    const buildPrior = (input: RapmProcessingInputs) => {
      const vec = input.rapmRawAdjPpp;
      return _.reduce(vec, (acc, n: number, i: number) => acc + n*input.playerPossPcts[i]!) || 0
    };
    const [ sigmaRapmOff, sigmaRapmDef ]  = [ buildPrior(offInputs), buildPrior(defInputs) ];
    const offPriorTotalDiff = teamOffAdj - sigmaRapmOff;
    const defPriorTotalDiff = teamDefAdj - sigmaRapmDef;
    const offPriorContrib = rapmOff - offUnbiasRapm;
    const defPriorContrib = rapmDef - defUnbiasRapm;

    const maybeAdaptiveWeight = rapmInfo?.preProcDiags?.adaptiveCorrelWeights?.[col] || 0.0;

    const detailedInfoPost = <ul>
      <li>We calculate a team adjustment (off=[<b>{offPriorTotalDiff.toFixed(2)}</b>] def=[<b>{defPriorTotalDiff.toFixed(2)}</b>]) to reduce/remove the
      the delta between total adjusted efficiency and RAPM (due to the regression factor): eg compare <em>observed</em> (off=[<b>{teamOffAdj.toFixed(2)}</b>] def=[<b>{teamDefAdj.toFixed(2)}</b>])
      vs <em>derived solely from RAPM</em> (off=[<b>{sigmaRapmOff.toFixed(2)}</b>] def=[<b>{sigmaRapmDef.toFixed(2)}</b>]).
      </li>
      <li>Then we calculate a player's contribution to this team total - currently this is a fraction of
      "Adj Rtg+": off=[<b>{offPrior.toFixed(2)}</b>] def=[<b>{defPrior.toFixed(2)}</b>]), ...
      </li>
      <li>
      ... chosen so that a minutes-weighted average of the ratings sums to the team value: off=[<b>{offPriorContrib.toFixed(2)}</b>], def=[<b>{defPriorContrib.toFixed(2)}</b>]
        <ul>
          <li>
          <em> (eg incorporating the % on floor [<b>{offPossPctStr}%</b>] (of [<b>{totalOffPoss}</b>] poss,
             this is an off=[<b>{(offPriorContrib*offPoss).toFixed(2)}</b>] def=[<b>{(defPriorContrib*defPoss).toFixed(2)}</b>]
            "slice" of the team total of off=[<b>{offPriorTotalDiff.toFixed(2)}</b>] def=[<b>{defPriorTotalDiff.toFixed(2)}</b>])</em>
          </li>
        </ul>
      </li>
    </ul>;

    const adaptiveWeight = ctx.priorInfo.strongWeight >= 0 ? ctx.priorInfo.strongWeight : maybeAdaptiveWeight;
    const rapmPriorOverrideInfo = ctx.priorInfo.strongWeight >= 0 ?
      <span> (hand-overwritten to <b>{(adaptiveWeight).toFixed(2)}</b>)</span> : null;

    const detailedInfoPre = <ul>
      <li>To combat the tendency of RAPM to over-share the contribution of the strongest players amongst their typical team-mates,
      we take the weighted average player correlation (see "Player correlation table" in the Global Diagnostics below) [<b>{(maybeAdaptiveWeight).toFixed(2)}</b>]{rapmPriorOverrideInfo}, ...
      </li>
      <li>... and use that % of the player's "Adj Rating+" ([<b>{(adaptiveWeight).toFixed(2)}</b>]*[<b>{(offPrior).toFixed(2)}</b>]) = [<b>{(adaptiveWeight*offPrior).toFixed(2)}</b>] as a prior in the RAPM calculation.
      </li>
      <ul>
        <li><i>(Currently we only do this for offense because ORtg is a much more reliable individual stat than DRtg)</i></li>
      </ul>
      <li>ie Off RAPM [<b>{offUnbiasRapm.toFixed(2)}</b>] = Raw RAPM [<b>{(offUnbiasRapm - adaptiveWeight*offPrior).toFixed(2)}</b>] + Prior [<b>{(adaptiveWeight*offPrior).toFixed(2)}</b>]</li>
    </ul>;

    const totalPrior = offPriorContrib - defPriorContrib;
    const totalRawRapm = offUnbiasRapm - defUnbiasRapm;

    //TODO: play around with player RAPM 
    const offMatrixInputs = rapmInfo.offInputs.solnMatrix.valueOf()[col];
    const [ offAdjPoss, defAdjPoss ] = RapmUtils.calcLineupOutputs(
      "adj_ppp", ctx.avgEfficiency, ctx.avgEfficiency, ctx, ctx.priorInfo.adaptiveCorrelWeights, false
    );
    const rapmDiag = offMatrixInputs.map((p, n) => { return {
      lineupId: n,
      lineupVal: offAdjPoss[n]!,
      matrixVal: p,
      contrib: p*offAdjPoss[n]!
    }; });
    const goodOnLineups = _.chain(rapmDiag).filter(d => d.matrixVal > 0 && d.contrib > 0).sortBy([d => -1*d.contrib]).value();
    const badOnLineups = _.chain(rapmDiag).filter(d => d.matrixVal > 0 && d.contrib < 0).sortBy([d => d.contrib]).value();
    const goodOffLineups = _.chain(rapmDiag).filter(d => d.matrixVal < 0 && d.contrib > 0).sortBy([d => -1*d.contrib]).value();
    const badOffLineups = _.chain(rapmDiag).filter(d => d.matrixVal < 0 && d.contrib < 0).sortBy([d => d.contrib]).value();
    const goodOnTotal = _.chain(goodOnLineups).sumBy(d => d.contrib).value().toFixed(2);
    const goodOnOther = _.chain(goodOnLineups).drop(5).sumBy(d => d.contrib).value().toFixed(2);
    const badOnTotal = _.chain(badOnLineups).sumBy(d => d.contrib).value().toFixed(2);
    const badOnOther = _.chain(badOnLineups).drop(5).sumBy(d => d.contrib).value().toFixed(2);
    const goodOffTotal = _.chain(goodOffLineups).sumBy(d => d.contrib).value().toFixed(2);
    const goodOffOther = _.chain(goodOffLineups).drop(5).sumBy(d => d.contrib).value().toFixed(2);
    const badOffTotal = _.chain(badOffLineups).sumBy(d => d.contrib).value().toFixed(2);
    const badOffOther = _.chain(badOffLineups).drop(5).sumBy(d => d.contrib).value().toFixed(2);
    const test = <span>
      Good On: {_.take(goodOnLineups, 5).map(d => d.contrib.toFixed(2)).join(" ; ")} Other=[{goodOnOther}, {goodOnLineups.length - 5}] Total=[{goodOnTotal}]<br/>
      Bad On: {_.take(badOnLineups, 5).map(d => d.contrib.toFixed(2)).join(" ; ")} Other=[{badOnOther}, {badOnLineups.length - 5}] Total=[{badOnTotal}]<br/>
      Good Off: {_.take(goodOffLineups, 5).map(d => d.contrib.toFixed(2)).join(" ; ")} Other=[{goodOffOther}, {goodOffLineups.length - 5}] Total=[{goodOffTotal}]<br/>
      Bad Off: {_.take(badOffLineups, 5).map(d => d.contrib.toFixed(2)).join(" ; ")} Other=[{badOffOther}, {badOffLineups.length - 5}] Total=[{badOffTotal}]<br/>
    </span>;

    return <span>
        <b>RAPM diagnostics for [{player.playerId}]:</b> adj_off=[<b>{rapmOff.toFixed(2)}</b>], adj_def=[<b>{rapmDef.toFixed(2)}</b>] =
        <ul>
          <li>RAPM contribution: off=[<b>{offUnbiasRapm.toFixed(2)}</b>], def=[<b>{defUnbiasRapm.toFixed(2)}</b>], total=[<b>{totalRawRapm.toFixed(2)}</b>]</li>
            {detailedInfoPre}
          <li>&nbsp;+ POST RAPM adjustment: off=[<b>{offPriorContrib.toFixed(2)}</b>], def=[<b>{defPriorContrib.toFixed(2)}</b>], total=[<b>{totalPrior.toFixed(2)}</b>]</li>
            {detailedInfoPost}
        </ul>
        {test}<br/>
        (<b>More player diagnostics to come...</b>)<br/>(<a href="#" onClick={(event) => { event.preventDefault(); gotoGlobalDiags() }}>Scroll to global RAPM diagnostics</a>)
      </span>;
    }
    catch (err) { //Temp issue during reprocessing
      return <span>Recalculating diags, pending {err.message}</span>;
    }
});

export default RapmPlayerDiagView;
