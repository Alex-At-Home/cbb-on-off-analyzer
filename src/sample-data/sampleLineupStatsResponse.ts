export const sampleLineupStatsResponse = { //TODO: need to change this...
   "took": 204,
   "responses": [
      {
         "took": 204,
         "timed_out": false,
         "_shards": {
            "total": 1,
            "successful": 1,
            "skipped": 0,
            "failed": 0
         },
         "hits": {
            "total": {
               "value": 318,
               "relation": "eq"
            },
            "max_score": null,
            "hits": []
         },
         "aggregations": {
            "lineups": {
               "doc_count_error_upper_bound": 0,
               "sum_other_doc_count": 0,
               "buckets": [
                 {
                    "key": "AaWiggins_AnCowan_DaMorsell_ErAyala_JaSmith",
                    "doc_count": 38,
                    "total_off_fga": {
                       "value": 167
                    },
                    "total_def_2prim_attempts": {
                       "value": 40
                    },
                    "total_def_drb": {
                       "value": 68
                    },
                    "total_off_2pmid_made": {
                       "value": 12
                    },
                    "total_off_2pmid_attempts": {
                       "value": 44
                    },
                    "total_def_fta": {
                       "value": 47
                    },
                    "total_def_2p_made": {
                       "value": 40
                    },
                    "off_2pmidr": {
                       "value": 0.2634730538922156
                    },
                    "off_ppp": {
                       "value": 111.22448979591837
                    },
                    "off_3pr": {
                       "value": 0.38922155688622756
                    },
                    "total_off_3p_attempts": {
                       "value": 65
                    },
                    "total_off_2p_made": {
                       "value": 52
                    },
                    "total_def_fga": {
                       "value": 158
                    },
                    "total_def_2p_attempts": {
                       "value": 92
                    },
                    "def_poss": {
                       "value": 189
                    },
                    "total_off_2prim_made": {
                       "value": 40
                    },
                    "def_to": {
                       "value": 0.14285714285714285
                    },
                    "def_adj_opp": {
                       "value": 95.56989795918368
                    },
                    "off_2primr": {
                       "value": 0.3473053892215569
                    },
                    "def_2primr": {
                       "value": 0.25316455696202533
                    },
                    "total_off_pts": {
                       "value": 218
                    },
                    "total_off_orb": {
                       "value": 39
                    },
                    "off_adj_opp": {
                       "value": 103.93650793650794
                    },
                    "def_ppp": {
                       "value": 90.47619047619048
                    },
                    "total_def_2pmid_attempts": {
                       "value": 52
                    },
                    "def_2pmidr": {
                       "value": 0.3291139240506329
                    },
                    "total_off_3p_made": {
                       "value": 18
                    },
                    "def_3pr": {
                       "value": 0.4177215189873418
                    },
                    "total_def_3p_made": {
                       "value": 19
                    },
                    "total_def_pts": {
                       "value": 171
                    },
                    "total_def_orb": {
                       "value": 19
                    },
                    "players_array": {
                       "hits": {
                          "total": {
                             "value": 38,
                             "relation": "eq"
                          },
                          "max_score": 4.760141,
                          "hits": [
                             {
                                "_index": "bigten_2019",
                                "_type": "_doc",
                                "_id": "k5mWzG4B8OZ7gWZVgjRk",
                                "_score": 4.760141,
                                "_source": {
                                   "players": [
                                      {
                                         "code": "AaWiggins",
                                         "id": "Wiggins, Aaron"
                                      },
                                      {
                                         "code": "AnCowan",
                                         "id": "Cowan, Anthony"
                                      },
                                      {
                                         "code": "DaMorsell",
                                         "id": "Morsell, Darryl"
                                      },
                                      {
                                         "code": "ErAyala",
                                         "id": "Ayala, Eric"
                                      },
                                      {
                                         "code": "JaSmith",
                                         "id": "Smith, Jalen"
                                      }
                                   ]
                                }
                             }
                          ]
                       }
                    },
                    "total_off_2p_attempts": {
                       "value": 102
                    },
                    "total_def_3p_attempts": {
                       "value": 66
                    },
                    "off_poss": {
                       "value": 196
                    },
                    "total_off_2prim_attempts": {
                       "value": 58
                    },
                    "total_def_2pmid_made": {
                       "value": 17
                    },
                    "total_def_2prim_made": {
                       "value": 23
                    },
                    "off_efg": {
                       "value": 0.47305389221556887
                    },
                    "off_to": {
                       "value": 0.14285714285714285
                    },
                    "total_off_drb": {
                       "value": 90
                    },
                    "total_off_fta": {
                       "value": 83
                    },
                    "def_efg": {
                       "value": 0.43354430379746833
                    },
                    "off_2p": {
                       "value": 0.5098039215686274
                    },
                    "def_2p": {
                       "value": 0.43478260869565216
                    },
                    "off_3p": {
                       "value": 0.27692307692307694
                    },
                    "def_3p": {
                       "value": 0.2878787878787879
                    },
                    "off_2prim": {
                       "value": 0.6896551724137931
                    },
                    "def_2prim": {
                       "value": 0.575
                    },
                    "off_2pmid": {
                       "value": 0.2727272727272727
                    },
                    "def_2pmid": {
                       "value": 0.3269230769230769
                    },
                    "off_ftr": {
                       "value": 0.49700598802395207
                    },
                    "def_ftr": {
                       "value": 0.2974683544303797
                    },
                    "off_orb": {
                       "value": 0.3644859813084112
                    },
                    "def_orb": {
                       "value": 0.1743119266055046
                    }
                 },
                 {
                    "key": "AaWiggins_AnCowan_DaMorsell_DoScott_JaSmith",
                    "doc_count": 27,
                    "total_off_fga": {
                       "value": 87
                    },
                    "total_def_2prim_attempts": {
                       "value": 27
                    },
                    "total_def_drb": {
                       "value": 43
                    },
                    "total_off_2pmid_made": {
                       "value": 4
                    },
                    "total_off_2pmid_attempts": {
                       "value": 27
                    },
                    "total_def_fta": {
                       "value": 24
                    },
                    "total_def_2p_made": {
                       "value": 22
                    },
                    "off_2pmidr": {
                       "value": 0.3103448275862069
                    },
                    "off_ppp": {
                       "value": 89.1891891891892
                    },
                    "off_3pr": {
                       "value": 0.3563218390804598
                    },
                    "total_off_3p_attempts": {
                       "value": 31
                    },
                    "total_off_2p_made": {
                       "value": 22
                    },
                    "total_def_fga": {
                       "value": 85
                    },
                    "total_def_2p_attempts": {
                       "value": 58
                    },
                    "def_poss": {
                       "value": 114
                    },
                    "total_off_2prim_made": {
                       "value": 18
                    },
                    "def_to": {
                       "value": 0.2719298245614035
                    },
                    "def_adj_opp": {
                       "value": 93.37927927927929
                    },
                    "off_2primr": {
                       "value": 0.3333333333333333
                    },
                    "def_2primr": {
                       "value": 0.3176470588235294
                    },
                    "total_off_pts": {
                       "value": 99
                    },
                    "total_off_orb": {
                       "value": 18
                    },
                    "off_adj_opp": {
                       "value": 106.32105263157895
                    },
                    "def_ppp": {
                       "value": 77.19298245614036
                    },
                    "total_def_2pmid_attempts": {
                       "value": 31
                    },
                    "def_2pmidr": {
                       "value": 0.36470588235294116
                    },
                    "total_off_3p_made": {
                       "value": 7
                    },
                    "def_3pr": {
                       "value": 0.3176470588235294
                    },
                    "total_def_3p_made": {
                       "value": 9
                    },
                    "total_def_pts": {
                       "value": 88
                    },
                    "total_def_orb": {
                       "value": 14
                    },
                    "players_array": {
                       "hits": {
                          "total": {
                             "value": 27,
                             "relation": "eq"
                          },
                          "max_score": 4.760141,
                          "hits": [
                             {
                                "_index": "bigten_2019",
                                "_type": "_doc",
                                "_id": "apmWzG4B8OZ7gWZVgTSi",
                                "_score": 4.760141,
                                "_source": {
                                   "players": [
                                      {
                                         "code": "AaWiggins",
                                         "id": "Wiggins, Aaron"
                                      },
                                      {
                                         "code": "AnCowan",
                                         "id": "Cowan, Anthony"
                                      },
                                      {
                                         "code": "DaMorsell",
                                         "id": "Morsell, Darryl"
                                      },
                                      {
                                         "code": "DoScott",
                                         "id": "Scott, Donta"
                                      },
                                      {
                                         "code": "JaSmith",
                                         "id": "Smith, Jalen"
                                      }
                                   ]
                                }
                             }
                          ]
                       }
                    },
                    "total_off_2p_attempts": {
                       "value": 56
                    },
                    "total_def_3p_attempts": {
                       "value": 27
                    },
                    "off_poss": {
                       "value": 111
                    },
                    "total_off_2prim_attempts": {
                       "value": 29
                    },
                    "total_def_2pmid_made": {
                       "value": 8
                    },
                    "total_def_2prim_made": {
                       "value": 14
                    },
                    "off_efg": {
                       "value": 0.3735632183908046
                    },
                    "off_to": {
                       "value": 0.18018018018018017
                    },
                    "total_off_drb": {
                       "value": 43
                    },
                    "total_off_fta": {
                       "value": 48
                    },
                    "def_efg": {
                       "value": 0.4176470588235294
                    },
                    "off_2p": {
                       "value": 0.39285714285714285
                    },
                    "def_2p": {
                       "value": 0.3793103448275862
                    },
                    "off_3p": {
                       "value": 0.22580645161290322
                    },
                    "def_3p": {
                       "value": 0.3333333333333333
                    },
                    "off_2prim": {
                       "value": 0.6206896551724138
                    },
                    "def_2prim": {
                       "value": 0.5185185185185185
                    },
                    "off_2pmid": {
                       "value": 0.14814814814814814
                    },
                    "def_2pmid": {
                       "value": 0.25806451612903225
                    },
                    "off_ftr": {
                       "value": 0.5517241379310345
                    },
                    "def_ftr": {
                       "value": 0.2823529411764706
                    },
                    "off_orb": {
                       "value": 0.29508196721311475
                    },
                    "def_orb": {
                       "value": 0.24561403508771928
                    }
                 },
                 {
                    "key": "AaWiggins_AnCowan_DoScott_ErAyala_JaSmith",
                    "doc_count": 24,
                    "total_off_fga": {
                       "value": 96
                    },
                    "total_def_2prim_attempts": {
                       "value": 28
                    },
                    "total_def_drb": {
                       "value": 36
                    },
                    "total_off_2pmid_made": {
                       "value": 10
                    },
                    "total_off_2pmid_attempts": {
                       "value": 25
                    },
                    "total_def_fta": {
                       "value": 27
                    },
                    "total_def_2p_made": {
                       "value": 17
                    },
                    "off_2pmidr": {
                       "value": 0.2604166666666667
                    },
                    "off_ppp": {
                       "value": 109.80392156862744
                    },
                    "off_3pr": {
                       "value": 0.4166666666666667
                    },
                    "total_off_3p_attempts": {
                       "value": 40
                    },
                    "total_off_2p_made": {
                       "value": 26
                    },
                    "total_def_fga": {
                       "value": 93
                    },
                    "total_def_2p_attempts": {
                       "value": 50
                    },
                    "def_poss": {
                       "value": 97
                    },
                    "total_off_2prim_made": {
                       "value": 16
                    },
                    "def_to": {
                       "value": 0.21649484536082475
                    },
                    "def_adj_opp": {
                       "value": 93.46372549019607
                    },
                    "off_2primr": {
                       "value": 0.3229166666666667
                    },
                    "def_2primr": {
                       "value": 0.3010752688172043
                    },
                    "total_off_pts": {
                       "value": 112
                    },
                    "total_off_orb": {
                       "value": 21
                    },
                    "off_adj_opp": {
                       "value": 104.70927835051546
                    },
                    "def_ppp": {
                       "value": 80.41237113402062
                    },
                    "total_def_2pmid_attempts": {
                       "value": 22
                    },
                    "def_2pmidr": {
                       "value": 0.23655913978494625
                    },
                    "total_off_3p_made": {
                       "value": 15
                    },
                    "def_3pr": {
                       "value": 0.46236559139784944
                    },
                    "total_def_3p_made": {
                       "value": 10
                    },
                    "total_def_pts": {
                       "value": 78
                    },
                    "total_def_orb": {
                       "value": 27
                    },
                    "players_array": {
                       "hits": {
                          "total": {
                             "value": 24,
                             "relation": "eq"
                          },
                          "max_score": 4.760141,
                          "hits": [
                             {
                                "_index": "bigten_2019",
                                "_type": "_doc",
                                "_id": "FJmWzG4B8OZ7gWZVjTbp",
                                "_score": 4.760141,
                                "_source": {
                                   "players": [
                                      {
                                         "code": "AaWiggins",
                                         "id": "Wiggins, Aaron"
                                      },
                                      {
                                         "code": "AnCowan",
                                         "id": "Cowan, Anthony"
                                      },
                                      {
                                         "code": "DoScott",
                                         "id": "Scott, Donta"
                                      },
                                      {
                                         "code": "ErAyala",
                                         "id": "Ayala, Eric"
                                      },
                                      {
                                         "code": "JaSmith",
                                         "id": "Smith, Jalen"
                                      }
                                   ]
                                }
                             }
                          ]
                       }
                    },
                    "total_off_2p_attempts": {
                       "value": 56
                    },
                    "total_def_3p_attempts": {
                       "value": 43
                    },
                    "off_poss": {
                       "value": 102
                    },
                    "total_off_2prim_attempts": {
                       "value": 31
                    },
                    "total_def_2pmid_made": {
                       "value": 3
                    },
                    "total_def_2prim_made": {
                       "value": 14
                    },
                    "off_efg": {
                       "value": 0.5052083333333334
                    },
                    "off_to": {
                       "value": 0.20588235294117646
                    },
                    "total_off_drb": {
                       "value": 43
                    },
                    "total_off_fta": {
                       "value": 19
                    },
                    "def_efg": {
                       "value": 0.34408602150537637
                    },
                    "off_2p": {
                       "value": 0.4642857142857143
                    },
                    "def_2p": {
                       "value": 0.34
                    },
                    "off_3p": {
                       "value": 0.375
                    },
                    "def_3p": {
                       "value": 0.23255813953488372
                    },
                    "off_2prim": {
                       "value": 0.5161290322580645
                    },
                    "def_2prim": {
                       "value": 0.5
                    },
                    "off_2pmid": {
                       "value": 0.4
                    },
                    "def_2pmid": {
                       "value": 0.13636363636363635
                    },
                    "off_ftr": {
                       "value": 0.19791666666666666
                    },
                    "def_ftr": {
                       "value": 0.2903225806451613
                    },
                    "off_orb": {
                       "value": 0.3684210526315789
                    },
                    "def_orb": {
                       "value": 0.38571428571428573
                    }
                 }
              ]
            }
         }
    }
  ]
}
