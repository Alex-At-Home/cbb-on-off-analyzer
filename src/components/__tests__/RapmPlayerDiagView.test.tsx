import renderer from 'react-test-renderer';
import React from 'react';
import RapmPlayerDiagView from '../RapmPlayerDiagView';

// Needed to build the data
import { sampleLineupStatsResponse } from "../../sample-data/sampleLineupStatsResponse";
import { LineupUtils } from "../../utils/LineupUtils";
import { RapmUtils } from "../../utils/RapmUtils";
import { semiRealRapmResults } from "../../utils/__tests__/RapmUtils.test";

describe("RapmPlayerDiagView", () => {

  const lineupReport = {
    lineups: sampleLineupStatsResponse.responses[0].aggregations.lineups.buckets,
    avgEff: 100.0,
    error_code: "test"
  };
  const globalRapmDiagRef = React.createRef<HTMLDivElement>();

  test("RapmPlayerDiagView - should create snapshot", () => {
    const [ offResults, defResults ] = RapmUtils.pickRidgeRegression(
      semiRealRapmResults.testOffWeights, semiRealRapmResults.testDefWeights, semiRealRapmResults.testContext, true
    );
    const onOffReport = LineupUtils.lineupToTeamReport(lineupReport);

    RapmUtils.injectRapmIntoPlayers(
      onOffReport.players || [], offResults, defResults, {}, semiRealRapmResults.testContext
    );
    const rapmInfo = {
      ctx: semiRealRapmResults.testContext,
      preProcDiags: RapmUtils.calcCollinearityDiag(
        semiRealRapmResults.testOffWeights, semiRealRapmResults.testContext
      ),
      noUnbiasWeightsDiags: RapmUtils.recalcNoUnbiasWeightingRapmForDiag(
        semiRealRapmResults.testOffWeights, semiRealRapmResults.testDefWeights,
        offResults, defResults, semiRealRapmResults.testContext
      ),
      offWeights: semiRealRapmResults.testOffWeights,
      defWeights: semiRealRapmResults.testDefWeights,
      offInputs: offResults,
      defInputs: defResults
    };

    const component = renderer.create(<RapmPlayerDiagView
      globalRef={globalRapmDiagRef}
      rapmInfo={rapmInfo}
      player={onOffReport.players?.[0] || {}}
    />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
