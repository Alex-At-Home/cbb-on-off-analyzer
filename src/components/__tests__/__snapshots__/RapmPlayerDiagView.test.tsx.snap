// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RapmPlayerDiagView RapmPlayerDiagView - should create snapshot 1`] = `
<span>
  <b>
    RAPM diagnostics for [
    Wiggins, Aaron
    ]:
  </b>
   adj_off=[
  <b>
    2.78
  </b>
  ], adj_def=[
  <b>
    -5.91
  </b>
  ]
  <ul>
    <li>
      RAPM contribution: off=[
      <b>
        2.98
      </b>
      ], def=[
      <b>
        -5.00
      </b>
      ], total=[
      <b>
        7.99
      </b>
      ]
    </li>
    <li>
      Priors contribution: off=[
      <b>
        -0.20
      </b>
      ], def=[
      <b>
        -0.90
      </b>
      ], total=[
      <b>
        0.70
      </b>
      ]
    </li>
    <ul>
      <li>
        Weighted down from the raw priors: off=[
        <b>
          1.59
        </b>
        ] def=[
        <b>
          -3.25
        </b>
        ]
      </li>
      <li>
        Calculated using players' % on floor, [
        <b>
          82
          %
        </b>
        ] of [
        <b>
          101
        </b>
        ] poss, and team's adj_off=[
        <b>
          10.0
        </b>
        ], adj_def=[
        <b>
          -20.0
        </b>
        ]
      </li>
    </ul>
  </ul>
  (
  <b>
    More to come...
  </b>
  )
</span>
`;
